src/ecmult_multi_impl.h

```diff
@@ -1,112 +1,192 @@
-            swap_idx = 2*root_idx + 1;
+
+        /* If the current node is larger than its largest child, stop at this level */
+        if (secp256k1_scalar_cmp_var(&sc[index], &sc[child_index]) > 0) {
+            break;
         }
-        /* If we're swapping, do it */
-        if (root_idx != swap_idx) {
-            SWAP(root_idx - 1, swap_idx - 1);
-            root_idx = swap_idx;
-        } else {
+
+        /* Move the larger child up, and recurse from its previous position */
+        heap->tree[node] = child_index;
+        node = child;
+    }
+
+    heap->tree[node] = index;
+}
+
+static void secp256k1_sift_up(secp256k1_scalar_heap *heap, size_t node, unsigned char index) {
+    size_t parent;
+    unsigned char parent_index;
+    const secp256k1_scalar *sc = heap->scalars;
+
+    while (node > 0) {
+        parent = (node - 1) >> 1;
+        parent_index = heap->tree[parent];
+
+        /* If the current node is not larger than its parent, stop at this level */
+        if (secp256k1_scalar_cmp_var(&sc[index], &sc[parent_index]) <= 0) {
             break;
         }
+
+        /* Move the parent down, and recurse from its previous position */
+        heap->tree[node] = parent_index;
+        node = parent;
     }
+
+    heap->tree[node] = index;
 }
-#undef SWAP
 
-static void secp256k1_heapify(const secp256k1_scalar *sc, unsigned char *idx, size_t n) {
-    size_t i;
-    for (i = n / 2; i > 0; i--) {
-        secp256k1_heap_siftdown(sc, idx, n, i);
+static void secp256k1_sift_floyd(secp256k1_scalar_heap *heap, size_t node, unsigned char index) {
+    unsigned char child_index, other_index;
+    size_t child, other, half_size = heap->size >> 1;
+    const secp256k1_scalar *sc = heap->scalars;
+
+    while (node < half_size) {
+        /* Initially assume the left child is the larger child */
+        child = (node << 1) + 1;
+        child_index = heap->tree[child];
+
+        /* If there is a right child, check whether it's larger than the left */
+        other = child + 1;
+        if (other < heap->size) {
+            other_index = heap->tree[other];
+            if (secp256k1_scalar_cmp_var(&sc[other_index], &sc[child_index]) > 0) {
+                child = other;
+                child_index = other_index;
+            }
+        }
+
+        /* Move the larger child up, and recurse from its previous position */
+        heap->tree[node] = child_index;
+        node = child;
     }
+
+    secp256k1_sift_up(heap, node, index);
 }
 
-static void secp256k1_heap_remove(const secp256k1_scalar *sc, unsigned char *idx, size_t *n) {
-    VERIFY_CHECK(*n > 0);
-    /* overwrite the root */
-    idx[0] = idx[*n - 1];
-    *n -= 1;
-    /* sift the new root into place */
-    secp256k1_heap_siftdown(sc, idx, *n, 1);
+SECP256K1_INLINE static void secp256k1_heapify(secp256k1_scalar_heap *heap) {
+    size_t root = heap->size >> 1;;
```

Comment:
Thanks.

This is far from the first time I've done this, I wonder what I'm doing to cause it.